src/compiler/diab.rs

```diff
@@ -0,0 +1,694 @@
+        .current_dir(cwd);
+
+    if log_enabled!(Trace) {
+        trace!("preprocess: {:?}", cmd);
+    }
+    run_input_output(cmd, None)
+}
+
+pub fn generate_compile_commands(
+    _path_transformer: &mut dist::PathTransformer,
+    executable: &Path,
+    parsed_args: &ParsedArguments,
+    cwd: &Path,
+    env_vars: &[(OsString, OsString)],
+) -> Result<(CompileCommand, Option<dist::CompileCommand>, Cacheable)> {
+    trace!("compile");
+
+    let out_file = match parsed_args.outputs.get("obj") {
+        Some(obj) => obj,
+        None => return Err("Missing object file output".into()),
+    };
+
+    let mut arguments: Vec<OsString> = vec![
+        "-c".into(),
+        parsed_args.input.clone().into(),
+        "-o".into(),
+        out_file.into(),
+    ];
+    arguments.extend(parsed_args.preprocessor_args.clone());
+    arguments.extend(parsed_args.common_args.clone());
+    let command = CompileCommand {
+        executable: executable.to_owned(),
+        arguments: arguments,
+        env_vars: env_vars.to_owned(),
+        cwd: cwd.to_owned(),
+    };
+
+    Ok((command, None, Cacheable::Yes))
+}
+
+pub struct ExpandAtArgs<'a> {
+    cwd: &'a Path,
+    stack: Vec<OsString>,
+}
+
+impl<'a> ExpandAtArgs<'a> {
+    pub fn new(cwd: &'a Path, args: &[OsString]) -> Self {
+        ExpandAtArgs {
+            stack: args.iter().rev().map(|a| a.to_owned()).collect(),
+            cwd: cwd,
+        }
+    }
+}
+
+impl<'a> Iterator for ExpandAtArgs<'a> {
+    type Item = OsString;
+
+    fn next(&mut self) -> Option<OsString> {
+        loop {
+            let arg = self.stack.pop()?;
+
+            // Just return non @ arguments
+            if !arg.starts_with("@") {
+                return Some(arg.into());
+            }
+
+            let arg = match arg.split_prefix("@") {
+                Some(arg) => arg,
+                None => return Some(arg),
+            };
+
+            // Return options that produce additional output and are not cacheable
+            if arg.starts_with("E=") || arg.starts_with("O=") || arg.starts_with("@") {
+                return Some(arg);
+            }
+
+            // According to diab [1], @file means:
+            // Read command line options from either a file or an environment
+            // variable. When -@name is encountered on the command line, the
+            // driver first looks for an environment variable with the given
+            // name and substitutes its value. If an environment variable is
+            // not found then the driver tries to open a file with given name
+            // and substitutes the contents of the file. If neither an
+            // environment variable or a file can be found, an error message
+            // is issued and the driver terminates.
+            //
+            // [1]: http://www.vxdev.com/docs/vx55man/diab5.0ppc/c-invoke.htm#3000619
+
+            if let Ok(env_arg) = env::var(&arg) {
```

Comment:
This isn't going to do the right thing--this will read the environment from the sccache server process. If you do:
```
sccache --start-server
FOO="-c foo.c -o foo.o" sccache diab -@FOO
```
The server process won't have the `FOO` environment variable. The server does have access to the client's environment variables at the time it calls `parse_arguments`:
https://github.com/mozilla/sccache/blob/3d9f934ae94759e6aef4b3c02bf0779df4fe2cc6/src/server.rs#L565

So you could change the signature of `parse_arguments` to pass them through, and change all the implementations to match.

Reply:
I don't know if it's workable for you, but an alternate tack would be to simply return `CannotCache` for `-@FOO` where "FOO" is not a file.

Reply:
Thanks for pointing the fail. At our place nobody uses the env var feature so I didn't get any reports for that. The tests are fine but I should have written a integration test to cover it completely.

Adding the environment to `Compiler::parse_arguments` would obviously affect *all* compiler implementations and add something unused there.

As commented before I've never seen anybody using this feature. When thinking about it I imagine totally confused colleagues who wonder why some piece of software behaves different depending on the machine built.

I'll follow your `CannotCache` suggestion. If anybody really needs this we can add it later.

Reply:
Done.